### PR TITLE
TW-141 Activities are idempotent

### PR DESCRIPTION
**Pull request by rachfop**
## What does this PR do?

Adds additional information about Activities being idempotent.

### CODE REVIEW

**Review by DJSanti — APPROVED** [review]

**Review by flossypurse — CHANGES_REQUESTED** [review]
I think this is actually relevant to explaining the core term "Activity" - so I actually want to see this added to the "What is an Activity" concept directly.

File: docs-src/app-dev-context/developing-activities.md [hunk 1]
```diff
@@ -10,3 +10,5 @@ tags:
 One of the primary things that Workflows do is orchestrate the execution of Activities.
 Activities are normal function/method executions that can interact with the world.
 For the Workflow to be able to execute the Activity, we must define the [Activity Definition](/concepts/what-is-an-activity-definition).
+
+Temporal recommends that Activities are idempotent. Activities are idempotent if multiple applications of that operation do not change the state of the system beyond the initial application.
```
Comment (cretz) [inline docs-src/app-dev-context/developing-activities.md:14]: More than recommends. They must be idempotent or things can go wrong.

File: docs/activities.md [hunk 2]
```diff
@@ -39,6 +39,22 @@ Activity Definitions are named and referenced in code by their <a class="tdlp" h
 
 <div class="tdiw"><div class="tditw"><p class="tdit">Activity Definition</p></div><div class="tdiiw"><img class="tdi" src="/diagrams/activity-definition.svg" alt="Activity Definition" /></div></div>
 
+#### Idempotency
+
+Temporal recommends that Activities are idempotent.
+
+Activities are idempotent if multiple applications of that operation do not change the state of the system beyond the initial application.
+
+Activities may be retried repeatedly, so you may need to use idempotency keys for critical side effects.
```
Comment (cretz) [inline docs/activities.md:48]: To clarify, they may also be executed more than once even if the retry policy says they shouldn't. So it's not just "retry".

File: docs/activities.md [hunk 3]
```diff
@@ -39,6 +39,22 @@ Activity Definitions are named and referenced in code by their <a class="tdlp" h
...
+
+Unlike determinism in a Workflow, which actually is a requirement, Temporal cannot enforce that your Activity is idempotent.
```
Comment (cretz) [inline docs/activities.md:50]: In many cases, Temporal cannot enforce that your workflow is deterministic. Activity idempotency and workflow determinism should be treated with the same importance IMO. No need for this contrast.

File: docs/activities.md [hunk 4]
```diff
@@ -39,6 +39,22 @@ Activity Definitions are named and referenced in code by their <a class="tdlp" h
...
+
+The lack of idempotency may affect the correctness of your application, but does not affect Temporal (in other words, it does not lead to a platform error).
+
+<!-- Temporal supports either at most once or at least once execution of Activities, and in the case of at least once activities must be idempotent -->
+
+In some cases, whether something is idempotent does not affect the correctness of an application. For example, if you have a monotonically-incrementing counter, you may not care that retries increment the counter because you don’t care about the actual value, only that the current value is greater than a previous value.
```
Comment (cretz) [inline docs/activities.md:56]: Then we may need to consider another word besides "idempotent" such as "reentrant" or we should call it "idempotent from a Temporal POV" or something. Whether something actually affects other parts of the user system is unimportant to Temporal's note that all activities may be run multiple times.

We don't care about the user's side, we just need them to know an activity can run multiple times. Maybe "idempotency" isn't the phrasing we should use. Maybe we should just tell people that all activities may run multiple times in an undetectable way and they should be prepared for that.

File: docs/concepts/what-is-an-activity.md [hunk 5]
```diff
@@ -11,7 +11,7 @@ tags:
 In day-to-day conversations, the term _Activity_ frequently denotes either an [Activity Definition](/concepts/what-is-an-activity-definition), an [Activity Type](/concepts/what-is-an-activity-type), or an [Activity Execution](/concepts/what-is-an-activity-execution).
 Temporal documentation aims to be explicit and differentiate between them.
 
-An Activity is a normal function or object method that executes a single, well-defined action (either short or long running), such as calling another service, transcoding a media file, or sending an email message.
+Temporal recommends that Activities are idempotent. Activities are idempotent if multiple applications of that operation do not change the state of the system beyond the initial application.
```
Comment (lorensr) [inline docs/concepts/what-is-an-activity.md:14]:
```suggestion
Temporal recommends that Activities be idempotent. Activities are idempotent if multiple [Activity Task Executions](/concepts/what-is-an-activity-task-execution) do not change the state of the system beyond the first Activity Task Execution.
```

File: docs/concepts/what-is-an-activity-definition.md [hunk 6]
```diff
@@ -19,6 +19,22 @@ Activity Definitions are named and referenced in code by their [Activity Type](/
 
 ![Activity Definition](/diagrams/activity-definition.svg)
 
+#### Idempotency
+
+Temporal recommends that Activities are idempotent.
+
+Activities are idempotent if multiple applications of that operation do not change the state of the system beyond the initial application.
```
Comment (lorensr) [inline docs/concepts/what-is-an-activity-definition.md:26]:
```suggestion
Temporal recommends that Activities be idempotent.

Activities are idempotent if multiple [Activity Task Executions](/concepts/what-is-an-activity-task-execution) do not change the state of the system beyond the first Activity Task Execution.
```

File: docs/concepts/what-is-an-activity.md (hunk 5 above)
Comment (lorensr) [reply docs/concepts/what-is-an-activity.md:14]: actually, since we don't have independent glossary pages anymore, I wonder whether we should remove this instance, since the same words will appear lower down on the activities page. https://docs.temporal.io/activities

also, I liked the deleted sentence:

An Activity is a normal function or object method that executes a single, well-defined action (either short or long running), such as calling another service, transcoding a media file, or sending an email message.

it gave me a practical idea of what an Activity is, which is harder to get if I read the page without it

File: docs-src/app-dev-context/developing-activities.md [hunk 7]
```diff
@@ -10,3 +10,5 @@ tags:
 One of the primary things that Workflows do is orchestrate the execution of Activities.
 Activities are normal function/method executions that can interact with the world.
 For the Workflow to be able to execute the Activity, we must define the [Activity Definition](/concepts/what-is-an-activity-definition).
+
+Temporal requires that Activities are idempotent. Activities are idempotent if multiple applications of that operation do not change the state of the system beyond the initial application.
```
Comment (flossypurse) [inline docs-src/app-dev-context/developing-activities.md:14]: Like to the Idempotency section in the concept?

File: docs-src/concepts/what-is-an-activity.md [hunk 8]
```diff
@@ -11,7 +11,7 @@ tags:
 In day-to-day conversations, the term _Activity_ frequently denotes either an [Activity Definition](/concepts/what-is-an-activity-definition), an [Activity Type](/concepts/what-is-an-activity-type), or an [Activity Execution](/concepts/what-is-an-activity-execution).
 Temporal documentation aims to be explicit and differentiate between them.
 
-An Activity is a normal function or object method that executes a single, well-defined action (either short or long running), such as calling another service, transcoding a media file, or sending an email message.
+Temporal recommends that Activities are idempotent. Activities are idempotent if multiple applications of that operation do not change the state of the system beyond the initial application.
```
Comment (flossypurse) [inline docs-src/concepts/what-is-an-activity.md:14]: Remove and revert this, as it is covered in the "Idempotency section" of the Activity Definition concept - and these files are back to back in the guide.

File: docs/activities.md [hunk 9]
```diff
@@ -30,15 +30,34 @@ The Event is added to the Workflow Execution's Event History.
 
 An Activity Definition is the code that defines the constraints of an <a class="tdlp" href="/tasks#activity-task-execution">Activity Task Execution<span class="tdlpiw"><img src="/img/link-preview-icon.svg" alt="Link preview icon" /></span><div class="tdlpc"><p class="tdlppt">What is an Activity Task Execution?</p><p class="tdlppd">An Activity Task Execution occurs when a Worker uses the context provided from the Activity Task and executes the Activity Definition.</p><p class="tdlplm"><a class="tdlplma" href="/tasks#activity-task-execution">Learn more</a></p></div></a>.
 
+One way to understand an Activity is to think of it as a normal function or object method that executes a single, well-defined action (either short or long-running), such as calling another service, transcoding a media file, or sending an email message.
```
Comment (flossypurse) [inline docs/activities.md:33]: Rewrite this to fit with the existing info in this section.
This is just a copy paste, and it doesn't mesh with the rest of the content here.

**Review by flossypurse — CHANGES_REQUESTED** [review]

File: docs-src/app-dev-context/developing-activities.md (hunk 7 above)
Comment (tomwheeler) [reply docs-src/app-dev-context/developing-activities.md:14]: I'll point out that this wording says that Temporal _requires_ Activities to be idempotent, but the wording in `docs-src/concepts/what-is-an-activity.md` says that Temporal _recommends_ them to be idempotent. 

Personally, I think that recommends (or perhaps "strongly recommends") is better wording, from the perspective that Temporal can do its job whether or not the Activity is idempotent (in contrast to whether or not a Workflow is deterministic, a violation of which can lead to failure). Others have tried to make the case that "required" is better. 

I will leave it up to the Docs team to decide which is more accurate, but I do suggest that Activity idempotence will confuse people if it's described as "recommended" in one place and "required" in another.

Comment (lorensr) [reply docs-src/app-dev-context/developing-activities.md:14]: It doesn't require, and there are times when idempotency isn't possible / supported by a downstream (possibly 3rd party) API

Comment (lorensr) [reply docs-src/app-dev-context/developing-activities.md:14]: 🤷‍♂️ I guess "in order for your application to be correct / always function as intended in the presence of failure, activities must be idempotent" can be shortened to "activities must be idempotent"

File: docs-src/app-dev-context/developing-activities.md [hunk 10]
```diff
@@ -8,5 +8,6 @@ tags:
 ---
 
 One of the primary things that Workflows do is orchestrate the execution of Activities.
-Activities are normal function/method executions that can interact with the world.
+Activities are normal function/method executions, each one intended to execute a single, well-defined action (either short or long-running) such as making a calculation, calling a third party API, or transcoding a media file.
+Activities can interact with world outside the Temporal Platform or use a Temporal Client to interact with a Cluster.
```
Comment (djmagee) [inline docs-src/app-dev-context/developing-activities.md:12]:
```suggestion
An Activity is a normal function or method execution that's intended to execute a single, well-defined action (either short or long-running), such as making a calculation, calling a third-party API, or transcoding a media file.
An Activity can interact with world outside the Temporal Platform or use a Temporal Client to interact with a Cluster.
```

File: docs-src/concepts/what-is-an-activity-definition.md [hunk 11]
```diff
@@ -19,6 +19,18 @@ Activity Definitions are named and referenced in code by their [Activity Type](/
 
 ![Activity Definition](/diagrams/activity-definition.svg)
 
+#### Idempotency
+
+Temporal recommends that Activities be idempotent.
+
+Activities are idempotent if multiple [Activity Task Executions](/concepts/what-is-an-activity-task-execution) do not change the state of the system beyond the first Activity Task Execution.
```
Comment (djmagee) [inline docs-src/concepts/what-is-an-activity-definition.md:26]:
```suggestion
An Activity is idempotent if multiple [Activity Task Executions](/concepts/what-is-an-activity-task-execution) do not change the state of the system beyond the first Activity Task Execution.
```

File: docs-src/concepts/what-is-an-activity-definition.md [hunk 12]
```diff
@@ -19,6 +19,18 @@ Activity Definitions are named and referenced in code by their [Activity Type](/
...
+
+Activities may be tried more than once even if the Retry Policy says they shouldn't. Because of that, you may need to use idempotency keys for critical side effects.
```
Comment (djmagee) [inline docs-src/concepts/what-is-an-activity-definition.md:28]:
```suggestion
An Activity can be tried more than once even if the Retry Policy says it shouldn't.
Because of that behavior, you might need to use idempotency keys for critical side effects.
```
Capitalize "Side Effects" and link to glossary term?

File: docs-src/concepts/what-is-an-activity-definition.md [hunk 13]
```diff
@@ -19,6 +19,18 @@ Activity Definitions are named and referenced in code by their [Activity Type](/
...
+
+The lack of idempotency may affect the correctness of your application, but does not affect Temporal (in other words, it does not lead to a platform error).
```
Comment (djmagee) [inline docs-src/concepts/what-is-an-activity-definition.md:30]:
```suggestion
The lack of idempotency might affect the correctness of your application but does not affect Temporal.
(In other words, lack of idempotency does not lead to a platform error.)
```
Should "Temporal" be "the Temporal Platform" in this instance?

File: docs-src/concepts/what-is-an-activity-definition.md [hunk 14]
```diff
@@ -19,6 +19,18 @@ Activity Definitions are named and referenced in code by their [Activity Type](/
...
+
+In some cases, whether something is idempotent does not affect the correctness of an application. For example, if you have a monotonically-incrementing counter, you may not care that retries increment the counter because you don’t care about the actual value, only that the current value is greater than a previous value.
```
Comment (djmagee) [inline docs-src/concepts/what-is-an-activity-definition.md:32]:
```suggestion
In some cases, whether something is idempotent does not affect the correctness of an application.
For example, if you have a monotonically incrementing counter, you might not care that retries increment the counter because you don’t care about the actual value, only that the current value is greater than a previous value.
```

**Review by djmagee — CHANGES_REQUESTED** [review]
Some edits and some questions.

File: docs/application-development/foundations.md [hunk 15]
```diff
@@ -1226,7 +1226,8 @@ Content is currently unavailable.
 ## Develop Activities
 
 One of the primary things that Workflows do is orchestrate the execution of Activities.
-Activities are normal function/method executions that can interact with the world.
+An Activity is a normal function or method execution that's intended to execute a single, well-defined action (either short or long-running), such as making a calculation, calling a third-party API, or transcoding a media file.
```
Comment (flossypurse) [inline docs/application-development/foundations.md:1229]:
```suggestion
An Activity is a normal function or method execution that's intended to execute a single, well-defined action (either short or long-running), such as querying a database, calling a third-party API, or transcoding a media file.
```

File: docs-src/app-dev-context/developing-activities.md [hunk 16]
```diff
@@ -8,5 +8,6 @@ tags:
 ---
 
 One of the primary things that Workflows do is orchestrate the execution of Activities.
-Activities are normal function/method executions that can interact with the world.
+An Activity is a normal function or method execution that's intended to execute a single, well-defined action (either short or long-running), such as making a calculation, calling a third-party API, or transcoding a media file.
```
Comment (flossypurse) [inline docs-src/app-dev-context/developing-activities.md:11]:
```suggestion
An Activity is a normal function or method execution that's intended to execute a single, well-defined action (either short or long-running), such as querying a database, calling a third-party API, or transcoding a media file.
```

File: docs/activities.md [hunk 17]
```diff
@@ -39,6 +39,21 @@ Activity Definitions are named and referenced in code by their <a class="tdlp" h
 
 <div class="tdiw"><div class="tditw"><p class="tdit">Activity Definition</p></div><div class="tdiiw"><img class="img_ev3q" src="/diagrams/activity-definition.svg" alt="Activity Definition" height="680" width="1000" /></div></div>
 
+#### Idempotency
+
+Temporal recommends that Activities be idempotent.
+
+An Activity is idempotent if multiple <a class="tdlp" href="/tasks#activity-task-execution">Activity Task Executions<span class="tdlpiw"><img src="/img/link-preview-icon.svg" alt="Link preview icon" /></span><span class="tdlpc"><span class="tdlppt">What is an Activity Task Execution?</span><br /><br /><span class="tdlppd">An Activity Task Execution occurs when a Worker uses the context provided from the Activity Task and executes the Activity Definition.</span><span class="tdlplm"><br /><br /><a class="tdlplma" href="/tasks#activity-task-execution">Learn more</a></span></span></a> do not change the state of the system beyond the first Activity Task Execution.
+
+An Activity can be tried more than once even if the Retry Policy says it shouldn't.
```
Comment (flossypurse) [inline docs/activities.md:48]: per @tomwheeler 

> This will inevitably lead the reader to ask "Why? Under what conditions would an Activity be tried more than once, contradicting what I put in the RetryPolicy?" I think the documentation will be stronger if that question is briefly answered right there where it's brought up, then linking to relevant documentation that explains it in more detail.

I agree - 
I think it is in some rare occurrences during a failover? I would definitely follow up on this point.

File: docs-src/typescript/clients.md [hunk 18]
```diff
@@ -41,13 +41,15 @@ import { example } from './workflows';
...
   // Connect to the default Server location (localhost:7233)
+  console.log(Date.now());
   const connection = await Connection.connect();
   // In production, pass options to configure TLS and other settings:
   // {
   //   address: 'foo.bar.tmprl.cloud',
   //   tls: {}
   // }
 
+  console.log(Date.now());
```
Comment (cretz) [inline docs-src/typescript/clients.md:52]: This and several other changes that merging this PR will make to the main branch don't seem related to the PR title/description

**Review by flossypurse — APPROVED** [review]
Nice.
I think this lgtm -

**Review by djmagee — APPROVED** [review]
